Chapter 20 web api rest (#18)
LaunchCodeEducation/csharp-web-dev-curriculum: content/web-api-rest/exercises/_index.md

}
```

In our case, the `CodingEvent` shape is just the properties and types (translated to portable [JSON types](https://json-schema.org/understanding-json-schema/reference/type.html) defined in a `CodingEvent` model class in `coding-events-api`.

```csharp {linenos=table}
public class CodingEvent {
   public int Id { get; set; }
   public string Title { get; set; }
   public string Description { get; set; }
   public DateTime Date { get; set; }
}
```

An example of a real `CodingEvent` JSON response to a `GET` request would look like this:

```bash {linenos=table}
{
   "id": 1,
   "title": "Consuming the Coding Events API With Postman",
   "description": "Learn how to use Postman to interact with the Coding Events API!",
   "date": "2020-07-24"
}
```

Notice how this JSON is just a representation of an instance of the `CodingEvent` model class. The data shape has been converted from a C# object representation to a JSON string representation so it can be transported over HTTP. Recall that we perform this 
conversion, or serialization, so that our API can output data in a portable format that is language-agnostic.

### The Endpoints

This branch of the API has four endpoints: 

> 1. [GET Coding Events](https://education.launchcode.org/csharp-web-dev-curriculum/web-api-rest/exercises/#get-coding-events)
> 1. [GET a Single Coding Event](https://education.launchcode.org/csharp-web-dev-curriculum/web-api-rest/exercises/#get-a-single-coding-event)
> 1. [Create a Coding Event](https://education.launchcode.org/csharp-web-dev-curriculum/web-api-rest/exercises/#create-a-coding-event)
> 1. [Delete a Coding Event](https://education.launchcode.org/csharp-web-dev-curriculum/web-api-rest/exercises/#delete-a-coding-event)

Reviewer: ```suggestion
1. [GET Coding Events](#get-coding-events)
1. [GET a Single Coding Event](#get-a-single-coding-event)
1. [Create a Coding Event](#create-a-coding-event)
1. [Delete a Coding Event](#delete-a-coding-event)
```

Deleting a `CodingEvent` resource is an operation on a single entity. Just like the endpoint for getting a single entity, this endpoint requires a `codingEventId` path variable. When a resource is deleted, a RESTful API should respond with a `204` status code. Similar to the `201` status, this code indicates a success with no response body or special headers. 

If you attempt to delete a resource that doesn't exist (with an incorrect `codingEventId`), then the endpoint will respond with an expected `404` status and message.

{{% notice green "Question" "rocket" %}}
Using our endpoint shorthand, how would we describe this action?

Some items to consider:

1. What is the HTTP request type being used?
1. What is the resource path being requested?
1. Is there a request body being sent? What is included in it?
1. If the request is successful, what information can we expect to be included in the response?
1. If the request contains an error, what information can we expect to be included in the response?
{{% /notice %}}

### Install Postman

Now that we've explored working with those endpoints, we're almost ready to start running the API and test sending those requests. 

> You'll need to [install Postman](https://education.launchcode.org/csharp-web-dev-curriculum/appendices/install-postman/) to work with this lesson's studio and practice running these requests.

Reviewer: ```suggestion
You'll need to [install Postman]({{< relref "../../../appendices/install-postman/_index.md" >}}) to work with this lesson's studio and practice running these requests.
```

---
title: "Exercises: Web APIs and REST"
date: 2022-12-15T09:16:07-06:00
draft: false
weight: 4
originalAuthor: John Woolbright # to be set by page creator
originalAuthorGitHub: jwoolbright23 # to be set by page creator
reviewer: Kimberly Horan # to be set by the page reviewer
reviewerGitHub: codinglikeagirl42 # to be set by the page reviewer
lastEditor: # update any time edits are made after review
lastEditorGitHub: # update any time edits are made after review
lastMod: 12/15/22 # UPDATE ANY TIME CHANGES ARE MADE
---

We've created a CodingEvents API application [here](https://github.com/LaunchCodeEducation/coding-events-api/tree/1-sqlite). Before we ask you to fork and clone this, let's first think conceptually about the kinds of requests a client application could make to this API.

## The API

> This branch of `coding-events-api` starts by only exposing a single resource, `CodingEvent`, and four endpoints for interacting with it.

Reviewer: ```suggestion
The 1-sqlite branch of `coding-events-api` starts by only exposing a single resource, `CodingEvent`, and four endpoints for interacting with it.
```

Author: Resolved this by removing the 1-sqlite branch altogether and having the learners work off of main.

Remember, an endpoint is made up of a path (to the resource) and a method (action to take on the 
resource). Because we only have one resource, each of our endpoints share a common entry-point path of `/api/events`. Request and response bodies are all in JSON, or more specifically, they have a `Content-Type` header value of `application/json`.

Making a `GET` request to the entry-point of a resource should return a representation of the state of the collection. In our case, this representation is a JSON array with `CodingEvent` elements:

```bash {linenos=table}
[
   CodingEvent { ... },
   ...
]
```

If the current state of the collection is empty, then we will just get back an empty JSON array:

```bash {linenos=table}
   []
```

{{% notice green "Question" "rocket" %}}
Using our endpoint shorthand, how would we describe this action?

Some items to consider:

1. What is the HTTP request type being used?
1. What is the resource path being requested?
1. Is there a request body being sent? What is included in it?
1. If the request is successful, what information can we expect to be included in the response?
{{% /notice %}}

> <!-- TODO: Need proper link here -->
> <!-- :ref:[Check your solutions](web-api-ex-1) -->

Reviewer: ```suggestion
{{% expand "Check Your Solution" %}}

{{% expand %}}
```

Reviewer: See exercises in classes chapter for reference of adding drop down Check your Solution and add solution.

Author: I do not think we need to add a solution to the exercises on this page after looking into it a bit further. The answers to the questions we ask are explained within the question itself. This is more of a check your understanding similar to a reading walkthrough.

Reviewer: Thanks for the explanation.

   "Date": "2020-10-31"
}
```

{{% notice blue "Note" "rocket" %}}
We only provide the user editable fields, not the unique `Id` which the API handles internally when saving to the database.
{{% /notice %}}

Recall that when a `POST` request is successful, the API should respond with the `201`, or *Created*, HTTP status code. As part of the `2XX`
HTTP success status codes, it indicates a particular type of successful response with a special header.

One of the REST conventions states that when an entity is created, the response should include both this status and the `Location` header that provides the URL of the new entity:

```bash {linenos=table}
Location=<server origin>/api/events/<new entity Id>
```

As an example:

```bash {linenos=table}
Location=http://localhost:5000/api/events/1
```

You can then issue a `GET` request to the `Location` header value and view the new entity. 

If the request fails because of a client error, then it will respond with a `400` status code and a message about what went wrong. In the case of 
`CodingEvent` entities, the following validation criteria must be met:

- `Title`: 10-100 characters
- `Description`: less than 1000 characters
 

> {{% notice blue "Question" "rocket" %}}

Reviewer: ```suggestion
{{% notice green "Question" "rocket" %}}
```
